// @wordpress/components/build/mobile/color-settings/picker-screen.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _native = require("@react-navigation/native");

var _components = require("@wordpress/components");

var _colorPicker = _interopRequireDefault(require("../../color-picker"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const PickerScreen = () => {
  const route = (0, _native.useRoute)();
  const navigation = (0, _native.useNavigation)();
  const {
    onShouldEnableInnerHandling,
    shouldEnableBottomSheetMaxHeight,
    onHandleClosingBottomSheet,
    isBottomSheetContentScrolling,
    shouldEnableBottomSheetScroll,
    onHandleHardwareButtonPress
  } = (0, _element.useContext)(_components.BottomSheetContext);
  const {
    setColor,
    currentValue,
    isGradientColor
  } = route.params;
  return (0, _element.useMemo)(() => {
    return (0, _element.createElement)(_colorPicker.default, {
      onShouldEnableInnerHandling: onShouldEnableInnerHandling,
      shouldEnableBottomSheetMaxHeight: shouldEnableBottomSheetMaxHeight,
      setColor: setColor,
      activeColor: currentValue,
      isGradientColor: isGradientColor,
      onNavigationBack: navigation.goBack,
      onHandleClosingBottomSheet: onHandleClosingBottomSheet,
      isBottomSheetContentScrolling: isBottomSheetContentScrolling,
      shouldEnableBottomSheetScroll: shouldEnableBottomSheetScroll,
      onHandleHardwareButtonPress: onHandleHardwareButtonPress
    });
  }, [setColor, currentValue, isGradientColor, onShouldEnableInnerHandling, shouldEnableBottomSheetMaxHeight, onHandleClosingBottomSheet, isBottomSheetContentScrolling, shouldEnableBottomSheetScroll, onHandleHardwareButtonPress]);
};

var _default = PickerScreen;
exports.default = _default;
