// @wordpress/block-editor/build/components/link-control/link-preview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LinkPreview;

var _element = require("@wordpress/element");

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

var _url = require("@wordpress/url");

var _icons = require("@wordpress/icons");

var _viewerSlot = require("./viewer-slot");

var _useRichUrlData = _interopRequireDefault(require("./use-rich-url-data"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function LinkPreview({
  value,
  onEditClick,
  hasRichPreviews = false
}) {
  // Avoid fetching if rich previews are not desired.
  const showRichPreviews = hasRichPreviews ? value === null || value === void 0 ? void 0 : value.url : null;
  const {
    richData,
    isFetching
  } = (0, _useRichUrlData.default)(showRichPreviews); // Rich data may be an empty object so test for that.

  const hasRichData = richData && Object.keys(richData).length;
  const displayURL = value && (0, _url.filterURLForDisplay)((0, _url.safeDecodeURI)(value.url), 16) || '';
  return (0, _element.createElement)("div", {
    "aria-label": (0, _i18n.__)('Currently selected'),
    "aria-selected": "true",
    className: (0, _classnames.default)('block-editor-link-control__search-item', {
      'is-current': true,
      'is-rich': hasRichData,
      'is-fetching': !!isFetching,
      'is-preview': true
    })
  }, (0, _element.createElement)("div", {
    className: "block-editor-link-control__search-item-top"
  }, (0, _element.createElement)("span", {
    className: "block-editor-link-control__search-item-header"
  }, (0, _element.createElement)("span", {
    className: (0, _classnames.default)('block-editor-link-control__search-item-icon', {
      'is-image': richData === null || richData === void 0 ? void 0 : richData.icon
    })
  }, richData !== null && richData !== void 0 && richData.icon ? (0, _element.createElement)("img", {
    src: richData === null || richData === void 0 ? void 0 : richData.icon,
    alt: ""
  }) : (0, _element.createElement)(_icons.Icon, {
    icon: _icons.globe
  })), (0, _element.createElement)("span", {
    className: "block-editor-link-control__search-item-details"
  }, (0, _element.createElement)(_components.ExternalLink, {
    className: "block-editor-link-control__search-item-title",
    href: value.url
  }, (richData === null || richData === void 0 ? void 0 : richData.title) || (value === null || value === void 0 ? void 0 : value.title) || displayURL), (value === null || value === void 0 ? void 0 : value.url) && (0, _element.createElement)("span", {
    className: "block-editor-link-control__search-item-info"
  }, displayURL))), (0, _element.createElement)(_components.Button, {
    variant: "secondary",
    onClick: () => onEditClick(),
    className: "block-editor-link-control__search-item-action"
  }, (0, _i18n.__)('Edit')), (0, _element.createElement)(_viewerSlot.ViewerSlot, {
    fillProps: value
  })), (hasRichData && ((richData === null || richData === void 0 ? void 0 : richData.image) || (richData === null || richData === void 0 ? void 0 : richData.description)) || isFetching) && (0, _element.createElement)("div", {
    className: "block-editor-link-control__search-item-bottom"
  }, ((richData === null || richData === void 0 ? void 0 : richData.image) || isFetching) && (0, _element.createElement)("div", {
    "aria-hidden": !(richData !== null && richData !== void 0 && richData.image),
    className: (0, _classnames.default)('block-editor-link-control__search-item-image', {
      'is-placeholder': !(richData !== null && richData !== void 0 && richData.image)
    })
  }, (richData === null || richData === void 0 ? void 0 : richData.image) && (0, _element.createElement)("img", {
    src: richData === null || richData === void 0 ? void 0 : richData.image,
    alt: ""
  })), ((richData === null || richData === void 0 ? void 0 : richData.description) || isFetching) && (0, _element.createElement)("div", {
    "aria-hidden": !(richData !== null && richData !== void 0 && richData.description),
    className: (0, _classnames.default)('block-editor-link-control__search-item-description', {
      'is-placeholder': !(richData !== null && richData !== void 0 && richData.description)
    })
  }, (richData === null || richData === void 0 ? void 0 : richData.description) && (0, _element.createElement)(_components.__experimentalText, {
    truncate: true,
    numberOfLines: "2"
  }, richData.description))));
}
