// @wordpress/block-editor/build/layouts/flow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _useSetting = _interopRequireDefault(require("../components/use-setting"));

var _utils = require("./utils");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var _default = {
  name: 'default',
  label: (0, _i18n.__)('Flow'),
  edit: function LayoutDefaultEdit({
    layout,
    onChange
  }) {
    const {
      wideSize,
      contentSize
    } = layout;
    const units = (0, _components.__experimentalUseCustomUnits)({
      availableUnits: (0, _useSetting.default)('spacing.units') || ['%', 'px', 'em', 'rem', 'vw']
    });
    return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)("div", {
      className: "block-editor-hooks__layout-controls"
    }, (0, _element.createElement)("div", {
      className: "block-editor-hooks__layout-controls-unit"
    }, (0, _element.createElement)(_components.__experimentalUnitControl, {
      label: (0, _i18n.__)('Content'),
      labelPosition: "top",
      __unstableInputWidth: "80px",
      value: contentSize || wideSize || '',
      onChange: nextWidth => {
        nextWidth = 0 > parseFloat(nextWidth) ? '0' : nextWidth;
        onChange({ ...layout,
          contentSize: nextWidth
        });
      },
      units: units
    }), (0, _element.createElement)(_icons.Icon, {
      icon: _icons.positionCenter
    })), (0, _element.createElement)("div", {
      className: "block-editor-hooks__layout-controls-unit"
    }, (0, _element.createElement)(_components.__experimentalUnitControl, {
      label: (0, _i18n.__)('Wide'),
      labelPosition: "top",
      __unstableInputWidth: "80px",
      value: wideSize || contentSize || '',
      onChange: nextWidth => {
        nextWidth = 0 > parseFloat(nextWidth) ? '0' : nextWidth;
        onChange({ ...layout,
          wideSize: nextWidth
        });
      },
      units: units
    }), (0, _element.createElement)(_icons.Icon, {
      icon: _icons.stretchWide
    }))), (0, _element.createElement)("div", {
      className: "block-editor-hooks__layout-controls-reset"
    }, (0, _element.createElement)(_components.Button, {
      variant: "secondary",
      isSmall: true,
      disabled: !contentSize && !wideSize,
      onClick: () => onChange({
        contentSize: undefined,
        wideSize: undefined,
        inherit: false
      })
    }, (0, _i18n.__)('Reset'))), (0, _element.createElement)("p", {
      className: "block-editor-hooks__layout-controls-helptext"
    }, (0, _i18n.__)('Customize the width for all elements that are assigned to the center or wide columns.')));
  },
  save: function DefaultLayoutStyle({
    selector,
    layout = {}
  }) {
    const {
      contentSize,
      wideSize
    } = layout;
    let style = !!contentSize || !!wideSize ? `
					${(0, _utils.appendSelectors)(selector, '> *')} {
						max-width: ${contentSize !== null && contentSize !== void 0 ? contentSize : wideSize};
						margin-left: auto !important;
						margin-right: auto !important;
					}
	
					${(0, _utils.appendSelectors)(selector, '> [data-align="wide"]')}  {
						max-width: ${wideSize !== null && wideSize !== void 0 ? wideSize : contentSize};
					}
	
					${(0, _utils.appendSelectors)(selector, '> [data-align="full"]')} {
						max-width: none;
					}
				` : '';
    style += `
			${(0, _utils.appendSelectors)(selector, '> [data-align="left"]')} {
				float: left;
				margin-right: 2em;
			}
	
			${(0, _utils.appendSelectors)(selector, '> [data-align="right"]')} {
				float: right;
				margin-left: 2em;
			}
		`;
    return (0, _element.createElement)("style", null, style);
  },

  getOrientation() {
    return 'vertical';
  },

  getAlignments(layout) {
    if (layout.alignments !== undefined) {
      return layout.alignments;
    }

    return layout.contentSize || layout.wideSize ? ['wide', 'full', 'left', 'center', 'right'] : ['left', 'center', 'right'];
  }

};
exports.default = _default;
