// @wordpress/block-editor/build/components/block-list/layout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useLayout = useLayout;
exports.LayoutStyle = LayoutStyle;
exports.LayoutProvider = exports.defaultLayout = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _layouts = require("../../layouts");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const defaultLayout = {
  type: 'default'
};
exports.defaultLayout = defaultLayout;
const Layout = (0, _element.createContext)(defaultLayout);
/**
 * Allows to define the layout.
 */

const LayoutProvider = Layout.Provider;
/**
 * React hook used to retrieve the layout config.
 */

exports.LayoutProvider = LayoutProvider;

function useLayout() {
  return (0, _element.useContext)(Layout);
}

function LayoutStyle({
  layout = {},
  ...props
}) {
  const layoutType = (0, _layouts.getLayoutType)(layout.type);

  if (layoutType) {
    return (0, _element.createElement)(layoutType.save, (0, _extends2.default)({
      layout: layout
    }, props));
  }

  return null;
}
