// @wordpress/components/build/custom-gradient-bar/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_MARKER = exports.KEYBOARD_CONTROL_POINT_VARIATION = exports.MINIMUM_SIGNIFICANT_MOVE = exports.MINIMUM_DISTANCE_BETWEEN_POINTS = exports.MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT = exports.MINIMUM_ABSOLUTE_LEFT_POSITION = exports.INSERT_POINT_WIDTH = exports.GRADIENT_MARKERS_WIDTH = exports.COLOR_POPOVER_PROPS = void 0;
const COLOR_POPOVER_PROPS = {
  className: 'components-custom-gradient-picker__color-picker-popover',
  position: 'top'
};
exports.COLOR_POPOVER_PROPS = COLOR_POPOVER_PROPS;
const GRADIENT_MARKERS_WIDTH = 18;
exports.GRADIENT_MARKERS_WIDTH = GRADIENT_MARKERS_WIDTH;
const INSERT_POINT_WIDTH = 23;
exports.INSERT_POINT_WIDTH = INSERT_POINT_WIDTH;
const MINIMUM_ABSOLUTE_LEFT_POSITION = 5;
exports.MINIMUM_ABSOLUTE_LEFT_POSITION = MINIMUM_ABSOLUTE_LEFT_POSITION;
const MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT = 10;
exports.MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT = MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT;
const MINIMUM_DISTANCE_BETWEEN_POINTS = 0;
exports.MINIMUM_DISTANCE_BETWEEN_POINTS = MINIMUM_DISTANCE_BETWEEN_POINTS;
const MINIMUM_SIGNIFICANT_MOVE = 5;
exports.MINIMUM_SIGNIFICANT_MOVE = MINIMUM_SIGNIFICANT_MOVE;
const KEYBOARD_CONTROL_POINT_VARIATION = MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT;
exports.KEYBOARD_CONTROL_POINT_VARIATION = KEYBOARD_CONTROL_POINT_VARIATION;
const MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_MARKER = (INSERT_POINT_WIDTH + GRADIENT_MARKERS_WIDTH) / 2;
exports.MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_MARKER = MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_MARKER;
