// @wordpress/components/build/text-control/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _cell = _interopRequireDefault(require("../mobile/bottom-sheet/cell"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function TextControl({
  label,
  hideLabelFromVision,
  value,
  help,
  className,
  instanceId,
  onChange,
  type = 'text',
  placeholder,
  ...props
}) {
  const id = `inspector-text-control-${instanceId}`;
  return (0, _element.createElement)(_cell.default, (0, _extends2.default)({
    label: label,
    hideLabelFromVision: hideLabelFromVision,
    id: id,
    help: help,
    className: className,
    type: type,
    value: value,
    onChangeValue: onChange,
    "aria-describedby": !!help ? id + '__help' : undefined,
    valuePlaceholder: placeholder
  }, props));
}

var _default = (0, _element.memo)(TextControl);

exports.default = _default;
