// @wordpress/blocks/build/api/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createBlock", {
  enumerable: true,
  get: function () {
    return _factory.createBlock;
  }
});
Object.defineProperty(exports, "createBlocksFromInnerBlocksTemplate", {
  enumerable: true,
  get: function () {
    return _factory.createBlocksFromInnerBlocksTemplate;
  }
});
Object.defineProperty(exports, "cloneBlock", {
  enumerable: true,
  get: function () {
    return _factory.cloneBlock;
  }
});
Object.defineProperty(exports, "__experimentalCloneSanitizedBlock", {
  enumerable: true,
  get: function () {
    return _factory.__experimentalCloneSanitizedBlock;
  }
});
Object.defineProperty(exports, "getPossibleBlockTransformations", {
  enumerable: true,
  get: function () {
    return _factory.getPossibleBlockTransformations;
  }
});
Object.defineProperty(exports, "switchToBlockType", {
  enumerable: true,
  get: function () {
    return _factory.switchToBlockType;
  }
});
Object.defineProperty(exports, "getBlockTransforms", {
  enumerable: true,
  get: function () {
    return _factory.getBlockTransforms;
  }
});
Object.defineProperty(exports, "findTransform", {
  enumerable: true,
  get: function () {
    return _factory.findTransform;
  }
});
Object.defineProperty(exports, "getBlockFromExample", {
  enumerable: true,
  get: function () {
    return _factory.getBlockFromExample;
  }
});
Object.defineProperty(exports, "parse", {
  enumerable: true,
  get: function () {
    return _parser.default;
  }
});
Object.defineProperty(exports, "getBlockAttributes", {
  enumerable: true,
  get: function () {
    return _getBlockAttributes.getBlockAttributes;
  }
});
Object.defineProperty(exports, "parseWithAttributeSchema", {
  enumerable: true,
  get: function () {
    return _getBlockAttributes.parseWithAttributeSchema;
  }
});
Object.defineProperty(exports, "pasteHandler", {
  enumerable: true,
  get: function () {
    return _rawHandling.pasteHandler;
  }
});
Object.defineProperty(exports, "rawHandler", {
  enumerable: true,
  get: function () {
    return _rawHandling.rawHandler;
  }
});
Object.defineProperty(exports, "getPhrasingContentSchema", {
  enumerable: true,
  get: function () {
    return _rawHandling.deprecatedGetPhrasingContentSchema;
  }
});
Object.defineProperty(exports, "serialize", {
  enumerable: true,
  get: function () {
    return _serializer.default;
  }
});
Object.defineProperty(exports, "getBlockContent", {
  enumerable: true,
  get: function () {
    return _serializer.getBlockInnerHTML;
  }
});
Object.defineProperty(exports, "getBlockDefaultClassName", {
  enumerable: true,
  get: function () {
    return _serializer.getBlockDefaultClassName;
  }
});
Object.defineProperty(exports, "getBlockMenuDefaultClassName", {
  enumerable: true,
  get: function () {
    return _serializer.getBlockMenuDefaultClassName;
  }
});
Object.defineProperty(exports, "getSaveElement", {
  enumerable: true,
  get: function () {
    return _serializer.getSaveElement;
  }
});
Object.defineProperty(exports, "getSaveContent", {
  enumerable: true,
  get: function () {
    return _serializer.getSaveContent;
  }
});
Object.defineProperty(exports, "__unstableGetBlockProps", {
  enumerable: true,
  get: function () {
    return _serializer.getBlockProps;
  }
});
Object.defineProperty(exports, "__unstableSerializeAndClean", {
  enumerable: true,
  get: function () {
    return _serializer.__unstableSerializeAndClean;
  }
});
Object.defineProperty(exports, "isValidBlockContent", {
  enumerable: true,
  get: function () {
    return _validation.isValidBlockContent;
  }
});
Object.defineProperty(exports, "getCategories", {
  enumerable: true,
  get: function () {
    return _categories.getCategories;
  }
});
Object.defineProperty(exports, "setCategories", {
  enumerable: true,
  get: function () {
    return _categories.setCategories;
  }
});
Object.defineProperty(exports, "updateCategory", {
  enumerable: true,
  get: function () {
    return _categories.updateCategory;
  }
});
Object.defineProperty(exports, "registerBlockType", {
  enumerable: true,
  get: function () {
    return _registration.registerBlockType;
  }
});
Object.defineProperty(exports, "registerBlockCollection", {
  enumerable: true,
  get: function () {
    return _registration.registerBlockCollection;
  }
});
Object.defineProperty(exports, "unregisterBlockType", {
  enumerable: true,
  get: function () {
    return _registration.unregisterBlockType;
  }
});
Object.defineProperty(exports, "setFreeformContentHandlerName", {
  enumerable: true,
  get: function () {
    return _registration.setFreeformContentHandlerName;
  }
});
Object.defineProperty(exports, "getFreeformContentHandlerName", {
  enumerable: true,
  get: function () {
    return _registration.getFreeformContentHandlerName;
  }
});
Object.defineProperty(exports, "setUnregisteredTypeHandlerName", {
  enumerable: true,
  get: function () {
    return _registration.setUnregisteredTypeHandlerName;
  }
});
Object.defineProperty(exports, "getUnregisteredTypeHandlerName", {
  enumerable: true,
  get: function () {
    return _registration.getUnregisteredTypeHandlerName;
  }
});
Object.defineProperty(exports, "setDefaultBlockName", {
  enumerable: true,
  get: function () {
    return _registration.setDefaultBlockName;
  }
});
Object.defineProperty(exports, "getDefaultBlockName", {
  enumerable: true,
  get: function () {
    return _registration.getDefaultBlockName;
  }
});
Object.defineProperty(exports, "setGroupingBlockName", {
  enumerable: true,
  get: function () {
    return _registration.setGroupingBlockName;
  }
});
Object.defineProperty(exports, "getGroupingBlockName", {
  enumerable: true,
  get: function () {
    return _registration.getGroupingBlockName;
  }
});
Object.defineProperty(exports, "getBlockType", {
  enumerable: true,
  get: function () {
    return _registration.getBlockType;
  }
});
Object.defineProperty(exports, "getBlockTypes", {
  enumerable: true,
  get: function () {
    return _registration.getBlockTypes;
  }
});
Object.defineProperty(exports, "getBlockSupport", {
  enumerable: true,
  get: function () {
    return _registration.getBlockSupport;
  }
});
Object.defineProperty(exports, "hasBlockSupport", {
  enumerable: true,
  get: function () {
    return _registration.hasBlockSupport;
  }
});
Object.defineProperty(exports, "getBlockVariations", {
  enumerable: true,
  get: function () {
    return _registration.getBlockVariations;
  }
});
Object.defineProperty(exports, "isReusableBlock", {
  enumerable: true,
  get: function () {
    return _registration.isReusableBlock;
  }
});
Object.defineProperty(exports, "isTemplatePart", {
  enumerable: true,
  get: function () {
    return _registration.isTemplatePart;
  }
});
Object.defineProperty(exports, "getChildBlockNames", {
  enumerable: true,
  get: function () {
    return _registration.getChildBlockNames;
  }
});
Object.defineProperty(exports, "hasChildBlocks", {
  enumerable: true,
  get: function () {
    return _registration.hasChildBlocks;
  }
});
Object.defineProperty(exports, "hasChildBlocksWithInserterSupport", {
  enumerable: true,
  get: function () {
    return _registration.hasChildBlocksWithInserterSupport;
  }
});
Object.defineProperty(exports, "unstable__bootstrapServerSideBlockDefinitions", {
  enumerable: true,
  get: function () {
    return _registration.unstable__bootstrapServerSideBlockDefinitions;
  }
});
Object.defineProperty(exports, "registerBlockStyle", {
  enumerable: true,
  get: function () {
    return _registration.registerBlockStyle;
  }
});
Object.defineProperty(exports, "unregisterBlockStyle", {
  enumerable: true,
  get: function () {
    return _registration.unregisterBlockStyle;
  }
});
Object.defineProperty(exports, "registerBlockVariation", {
  enumerable: true,
  get: function () {
    return _registration.registerBlockVariation;
  }
});
Object.defineProperty(exports, "unregisterBlockVariation", {
  enumerable: true,
  get: function () {
    return _registration.unregisterBlockVariation;
  }
});
Object.defineProperty(exports, "isUnmodifiedDefaultBlock", {
  enumerable: true,
  get: function () {
    return _utils.isUnmodifiedDefaultBlock;
  }
});
Object.defineProperty(exports, "normalizeIconObject", {
  enumerable: true,
  get: function () {
    return _utils.normalizeIconObject;
  }
});
Object.defineProperty(exports, "isValidIcon", {
  enumerable: true,
  get: function () {
    return _utils.isValidIcon;
  }
});
Object.defineProperty(exports, "__experimentalGetBlockLabel", {
  enumerable: true,
  get: function () {
    return _utils.getBlockLabel;
  }
});
Object.defineProperty(exports, "__experimentalGetAccessibleBlockLabel", {
  enumerable: true,
  get: function () {
    return _utils.getAccessibleBlockLabel;
  }
});
Object.defineProperty(exports, "__experimentalSanitizeBlockAttributes", {
  enumerable: true,
  get: function () {
    return _utils.__experimentalSanitizeBlockAttributes;
  }
});
Object.defineProperty(exports, "__experimentalGetBlockAttributesNamesByRole", {
  enumerable: true,
  get: function () {
    return _utils.__experimentalGetBlockAttributesNamesByRole;
  }
});
Object.defineProperty(exports, "doBlocksMatchTemplate", {
  enumerable: true,
  get: function () {
    return _templates.doBlocksMatchTemplate;
  }
});
Object.defineProperty(exports, "synchronizeBlocksWithTemplate", {
  enumerable: true,
  get: function () {
    return _templates.synchronizeBlocksWithTemplate;
  }
});
Object.defineProperty(exports, "children", {
  enumerable: true,
  get: function () {
    return _children.default;
  }
});
Object.defineProperty(exports, "node", {
  enumerable: true,
  get: function () {
    return _node.default;
  }
});
Object.defineProperty(exports, "__EXPERIMENTAL_STYLE_PROPERTY", {
  enumerable: true,
  get: function () {
    return _constants.__EXPERIMENTAL_STYLE_PROPERTY;
  }
});
Object.defineProperty(exports, "__EXPERIMENTAL_ELEMENTS", {
  enumerable: true,
  get: function () {
    return _constants.__EXPERIMENTAL_ELEMENTS;
  }
});

var _factory = require("./factory");

var _parser = _interopRequireDefault(require("./parser"));

var _getBlockAttributes = require("./parser/get-block-attributes");

var _rawHandling = require("./raw-handling");

var _serializer = _interopRequireWildcard(require("./serializer"));

var _validation = require("./validation");

var _categories = require("./categories");

var _registration = require("./registration");

var _utils = require("./utils");

var _templates = require("./templates");

var _children = _interopRequireDefault(require("./children"));

var _node = _interopRequireDefault(require("./node"));

var _constants = require("./constants");
