// @wordpress/components/build/box-control/linked-button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LinkedButton;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _icons = require("@wordpress/icons");

var _i18n = require("@wordpress/i18n");

var _button = _interopRequireDefault(require("../button"));

var _tooltip = _interopRequireDefault(require("../tooltip"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function LinkedButton({
  isLinked,
  ...props
}) {
  const label = isLinked ? (0, _i18n.__)('Unlink Sides') : (0, _i18n.__)('Link Sides');
  return (0, _element.createElement)(_tooltip.default, {
    text: label
  }, (0, _element.createElement)("span", null, (0, _element.createElement)(_button.default, (0, _extends2.default)({}, props, {
    className: "component-box-control__linked-button",
    variant: isLinked ? 'primary' : 'secondary',
    isSmall: true,
    icon: isLinked ? _icons.link : _icons.linkOff,
    iconSize: 16,
    "aria-label": label
  }))));
}
