// @wordpress/components/build/mobile/link-settings/link-settings-navigation.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _bottomSheet = _interopRequireDefault(require("../bottom-sheet"));

var _linkSettingsScreen = _interopRequireDefault(require("./link-settings-screen"));

var _linkPickerScreen = _interopRequireDefault(require("../link-picker/link-picker-screen"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const linkSettingsScreens = {
  settings: 'LinkSettingsScreen',
  linkPicker: 'linkPicker'
};

function LinkSettingsNavigation(props) {
  if (!props.withBottomSheet) {
    return (0, _element.createElement)(_linkSettingsScreen.default, props);
  }

  return (0, _element.createElement)(_bottomSheet.default, {
    isVisible: props.isVisible,
    onClose: props.onClose,
    onDismiss: props.onDismiss,
    hideHeader: true,
    hasNavigation: true
  }, (0, _element.createElement)(_bottomSheet.default.NavigationContainer, {
    animate: true,
    main: true
  }, (0, _element.createElement)(_bottomSheet.default.NavigationScreen, {
    name: linkSettingsScreens.settings
  }, (0, _element.createElement)(_linkSettingsScreen.default, (0, _extends2.default)({}, props, {
    withBottomSheet: true
  }))), (0, _element.createElement)(_bottomSheet.default.NavigationScreen, {
    name: linkSettingsScreens.linkPicker,
    isScrollable: true,
    fullScreen: true
  }, (0, _element.createElement)(_linkPickerScreen.default, {
    returnScreenName: linkSettingsScreens.settings
  }))));
}

var _default = (0, _element.memo)(LinkSettingsNavigation);

exports.default = _default;
