// @wordpress/components/build/utils/config-values.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _space = require("../ui/utils/space");

var _colorsValues = require("./colors-values");

/**
 * Internal dependencies
 */
const CONTROL_HEIGHT = '36px';
const CONTROL_PADDING_X = '12px';
const CONTROL_PROPS = {
  controlSurfaceColor: _colorsValues.COLORS.white,
  controlTextActiveColor: _colorsValues.COLORS.ui.theme,
  controlPaddingX: CONTROL_PADDING_X,
  controlPaddingXLarge: `calc(${CONTROL_PADDING_X} * 1.3334)`,
  controlPaddingXSmall: `calc(${CONTROL_PADDING_X} / 1.3334)`,
  controlBackgroundColor: _colorsValues.COLORS.white,
  controlBorderRadius: '2px',
  controlBorderColor: _colorsValues.COLORS.gray[700],
  controlBoxShadow: 'transparent',
  controlBorderColorHover: _colorsValues.COLORS.gray[700],
  controlBoxShadowFocus: `0 0 0 0.5px ${_colorsValues.COLORS.admin.theme}`,
  controlDestructiveBorderColor: _colorsValues.COLORS.alert.red,
  controlHeight: CONTROL_HEIGHT,
  controlHeightXSmall: `calc( ${CONTROL_HEIGHT} * 0.6 )`,
  controlHeightSmall: `calc( ${CONTROL_HEIGHT} * 0.8 )`,
  controlHeightLarge: `calc( ${CONTROL_HEIGHT} * 1.2 )`,
  controlHeightXLarge: `calc( ${CONTROL_HEIGHT} * 1.4 )`
};
const SEGMENTED_CONTROL_PROPS = {
  segmentedControlBackgroundColor: CONTROL_PROPS.controlBackgroundColor,
  segmentedControlBorderColor: _colorsValues.COLORS.ui.border,
  segmentedControlBackdropBackgroundColor: CONTROL_PROPS.controlSurfaceColor,
  segmentedControlBackdropBorderColor: _colorsValues.COLORS.ui.border,
  segmentedControlBackdropBoxShadow: 'transparent',
  segmentedControlButtonColorActive: CONTROL_PROPS.controlBackgroundColor
};
var _default = { ...CONTROL_PROPS,
  ...SEGMENTED_CONTROL_PROPS,
  colorDivider: 'rgba(0, 0, 0, 0.1)',
  colorScrollbarThumb: 'rgba(0, 0, 0, 0.2)',
  colorScrollbarThumbHover: 'rgba(0, 0, 0, 0.5)',
  colorScrollbarTrack: 'rgba(0, 0, 0, 0.04)',
  elevationIntensity: 1,
  radiusBlockUi: '2px',
  borderWidth: '1px',
  borderWidthFocus: '1.5px',
  borderWidthTab: '4px',
  spinnerSize: '18px',
  fontSize: '13px',
  fontSizeH1: 'calc(2.44 * 13px)',
  fontSizeH2: 'calc(1.95 * 13px)',
  fontSizeH3: 'calc(1.56 * 13px)',
  fontSizeH4: 'calc(1.25 * 13px)',
  fontSizeH5: '13px',
  fontSizeH6: 'calc(0.8 * 13px)',
  fontSizeInputMobile: '16px',
  fontSizeMobile: '15px',
  fontSizeSmall: 'calc(0.92 * 13px)',
  fontSizeXSmall: 'calc(0.75 * 13px)',
  fontLineHeightBase: '1.2',
  fontWeight: 'normal',
  fontWeightHeading: '600',
  gridBase: '4px',
  cardBorderRadius: '2px',
  cardPaddingXSmall: `${(0, _space.space)(2)}`,
  cardPaddingSmall: `${(0, _space.space)(4)}`,
  cardPaddingMedium: `${(0, _space.space)(4)} ${(0, _space.space)(6)}`,
  cardPaddingLarge: `${(0, _space.space)(6)} ${(0, _space.space)(8)}`,
  surfaceBackgroundColor: _colorsValues.COLORS.white,
  surfaceBackgroundSubtleColor: '#F3F3F3',
  surfaceBackgroundTintColor: '#F5F5F5',
  surfaceBorderColor: 'rgba(0, 0, 0, 0.1)',
  surfaceBorderBoldColor: 'rgba(0, 0, 0, 0.15)',
  surfaceBorderSubtleColor: 'rgba(0, 0, 0, 0.05)',
  surfaceBackgroundTertiaryColor: _colorsValues.COLORS.white,
  surfaceColor: _colorsValues.COLORS.white,
  transitionDuration: '200ms',
  transitionDurationFast: '160ms',
  transitionDurationFaster: '120ms',
  transitionDurationFastest: '100ms',
  transitionTimingFunction: 'cubic-bezier(0.08, 0.52, 0.52, 1)',
  transitionTimingFunctionControl: 'cubic-bezier(0.12, 0.8, 0.32, 1)'
};
exports.default = _default;
