// @wordpress/dom/build/dom/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "computeCaretRect", {
  enumerable: true,
  get: function () {
    return _computeCaretRect.default;
  }
});
Object.defineProperty(exports, "documentHasTextSelection", {
  enumerable: true,
  get: function () {
    return _documentHasTextSelection.default;
  }
});
Object.defineProperty(exports, "documentHasUncollapsedSelection", {
  enumerable: true,
  get: function () {
    return _documentHasUncollapsedSelection.default;
  }
});
Object.defineProperty(exports, "documentHasSelection", {
  enumerable: true,
  get: function () {
    return _documentHasSelection.default;
  }
});
Object.defineProperty(exports, "getRectangleFromRange", {
  enumerable: true,
  get: function () {
    return _getRectangleFromRange.default;
  }
});
Object.defineProperty(exports, "getScrollContainer", {
  enumerable: true,
  get: function () {
    return _getScrollContainer.default;
  }
});
Object.defineProperty(exports, "getOffsetParent", {
  enumerable: true,
  get: function () {
    return _getOffsetParent.default;
  }
});
Object.defineProperty(exports, "isEntirelySelected", {
  enumerable: true,
  get: function () {
    return _isEntirelySelected.default;
  }
});
Object.defineProperty(exports, "isHorizontalEdge", {
  enumerable: true,
  get: function () {
    return _isHorizontalEdge.default;
  }
});
Object.defineProperty(exports, "isNumberInput", {
  enumerable: true,
  get: function () {
    return _isNumberInput.default;
  }
});
Object.defineProperty(exports, "isTextField", {
  enumerable: true,
  get: function () {
    return _isTextField.default;
  }
});
Object.defineProperty(exports, "isVerticalEdge", {
  enumerable: true,
  get: function () {
    return _isVerticalEdge.default;
  }
});
Object.defineProperty(exports, "placeCaretAtHorizontalEdge", {
  enumerable: true,
  get: function () {
    return _placeCaretAtHorizontalEdge.default;
  }
});
Object.defineProperty(exports, "placeCaretAtVerticalEdge", {
  enumerable: true,
  get: function () {
    return _placeCaretAtVerticalEdge.default;
  }
});
Object.defineProperty(exports, "replace", {
  enumerable: true,
  get: function () {
    return _replace.default;
  }
});
Object.defineProperty(exports, "remove", {
  enumerable: true,
  get: function () {
    return _remove.default;
  }
});
Object.defineProperty(exports, "insertAfter", {
  enumerable: true,
  get: function () {
    return _insertAfter.default;
  }
});
Object.defineProperty(exports, "unwrap", {
  enumerable: true,
  get: function () {
    return _unwrap.default;
  }
});
Object.defineProperty(exports, "replaceTag", {
  enumerable: true,
  get: function () {
    return _replaceTag.default;
  }
});
Object.defineProperty(exports, "wrap", {
  enumerable: true,
  get: function () {
    return _wrap.default;
  }
});
Object.defineProperty(exports, "__unstableStripHTML", {
  enumerable: true,
  get: function () {
    return _stripHtml.default;
  }
});
Object.defineProperty(exports, "isEmpty", {
  enumerable: true,
  get: function () {
    return _isEmpty.default;
  }
});
Object.defineProperty(exports, "removeInvalidHTML", {
  enumerable: true,
  get: function () {
    return _removeInvalidHtml.default;
  }
});
Object.defineProperty(exports, "isRTL", {
  enumerable: true,
  get: function () {
    return _isRtl.default;
  }
});
Object.defineProperty(exports, "safeHTML", {
  enumerable: true,
  get: function () {
    return _safeHtml.default;
  }
});

var _computeCaretRect = _interopRequireDefault(require("./compute-caret-rect"));

var _documentHasTextSelection = _interopRequireDefault(require("./document-has-text-selection"));

var _documentHasUncollapsedSelection = _interopRequireDefault(require("./document-has-uncollapsed-selection"));

var _documentHasSelection = _interopRequireDefault(require("./document-has-selection"));

var _getRectangleFromRange = _interopRequireDefault(require("./get-rectangle-from-range"));

var _getScrollContainer = _interopRequireDefault(require("./get-scroll-container"));

var _getOffsetParent = _interopRequireDefault(require("./get-offset-parent"));

var _isEntirelySelected = _interopRequireDefault(require("./is-entirely-selected"));

var _isHorizontalEdge = _interopRequireDefault(require("./is-horizontal-edge"));

var _isNumberInput = _interopRequireDefault(require("./is-number-input"));

var _isTextField = _interopRequireDefault(require("./is-text-field"));

var _isVerticalEdge = _interopRequireDefault(require("./is-vertical-edge"));

var _placeCaretAtHorizontalEdge = _interopRequireDefault(require("./place-caret-at-horizontal-edge"));

var _placeCaretAtVerticalEdge = _interopRequireDefault(require("./place-caret-at-vertical-edge"));

var _replace = _interopRequireDefault(require("./replace"));

var _remove = _interopRequireDefault(require("./remove"));

var _insertAfter = _interopRequireDefault(require("./insert-after"));

var _unwrap = _interopRequireDefault(require("./unwrap"));

var _replaceTag = _interopRequireDefault(require("./replace-tag"));

var _wrap = _interopRequireDefault(require("./wrap"));

var _stripHtml = _interopRequireDefault(require("./strip-html"));

var _isEmpty = _interopRequireDefault(require("./is-empty"));

var _removeInvalidHtml = _interopRequireDefault(require("./remove-invalid-html"));

var _isRtl = _interopRequireDefault(require("./is-rtl"));

var _safeHtml = _interopRequireDefault(require("./safe-html"));
