// @wordpress/block-editor/build/hooks/border-radius.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BorderRadiusEdit = BorderRadiusEdit;

var _element = require("@wordpress/element");

var _borderRadiusControl = _interopRequireDefault(require("../components/border-radius-control"));

var _utils = require("./utils");

/**
 * Internal dependencies
 */

/**
 * Inspector control panel containing the border radius related configuration.
 *
 * @param {Object} props Block properties.
 *
 * @return {WPElement} Border radius edit element.
 */
function BorderRadiusEdit(props) {
  var _style$border;

  const {
    attributes: {
      style
    },
    setAttributes
  } = props;

  const onChange = newRadius => {
    let newStyle = { ...style,
      border: { ...(style === null || style === void 0 ? void 0 : style.border),
        radius: newRadius
      }
    };

    if (newRadius === undefined || newRadius === '') {
      newStyle = (0, _utils.cleanEmptyObject)(newStyle);
    }

    setAttributes({
      style: newStyle
    });
  };

  return (0, _element.createElement)(_borderRadiusControl.default, {
    values: style === null || style === void 0 ? void 0 : (_style$border = style.border) === null || _style$border === void 0 ? void 0 : _style$border.radius,
    onChange: onChange
  });
}
