// @wordpress/block-editor/build/components/duotone-control/duotone-picker-popover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

/**
 * WordPress dependencies
 */
function DuotonePickerPopover({
  value,
  onChange,
  onToggle,
  duotonePalette,
  colorPalette,
  disableCustomColors,
  disableCustomDuotone
}) {
  return (0, _element.createElement)(_components.Popover, {
    className: "block-editor-duotone-control__popover",
    headerTitle: (0, _i18n.__)('Duotone'),
    onFocusOutside: onToggle
  }, (0, _element.createElement)(_components.MenuGroup, {
    label: (0, _i18n.__)('Duotone')
  }, (0, _element.createElement)(_components.DuotonePicker, {
    colorPalette: colorPalette,
    duotonePalette: duotonePalette,
    disableCustomColors: disableCustomColors,
    disableCustomDuotone: disableCustomDuotone,
    value: value,
    onChange: onChange
  })));
}

var _default = DuotonePickerPopover;
exports.default = _default;
