// @wordpress/block-editor/build/components/block-card/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _deprecated = _interopRequireDefault(require("@wordpress/deprecated"));

var _blockIcon = _interopRequireDefault(require("../block-icon"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function BlockCard({
  title,
  icon,
  description,
  blockType
}) {
  if (blockType) {
    (0, _deprecated.default)('`blockType` property in `BlockCard component`', {
      since: '5.7',
      alternative: '`title, icon and description` properties'
    });
    ({
      title,
      icon,
      description
    } = blockType);
  }

  return (0, _element.createElement)("div", {
    className: "block-editor-block-card"
  }, (0, _element.createElement)(_blockIcon.default, {
    icon: icon,
    showColors: true
  }), (0, _element.createElement)("div", {
    className: "block-editor-block-card__content"
  }, (0, _element.createElement)("h2", {
    className: "block-editor-block-card__title"
  }, title), (0, _element.createElement)("span", {
    className: "block-editor-block-card__description"
  }, description)));
}

var _default = BlockCard;
exports.default = _default;
