// @wordpress/block-editor/build/components/inserter-listbox/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "InserterListboxGroup", {
  enumerable: true,
  get: function () {
    return _group.default;
  }
});
Object.defineProperty(exports, "InserterListboxRow", {
  enumerable: true,
  get: function () {
    return _row.default;
  }
});
Object.defineProperty(exports, "InserterListboxItem", {
  enumerable: true,
  get: function () {
    return _item.default;
  }
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _context = _interopRequireDefault(require("./context"));

var _group = _interopRequireDefault(require("./group"));

var _row = _interopRequireDefault(require("./row"));

var _item = _interopRequireDefault(require("./item"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function InserterListbox({
  children
}) {
  const compositeState = (0, _components.__unstableUseCompositeState)({
    shift: true,
    wrap: 'horizontal'
  });
  return (0, _element.createElement)(_context.default.Provider, {
    value: compositeState
  }, children);
}

var _default = InserterListbox;
exports.default = _default;
