// @wordpress/block-editor/build/components/inserter/block-types-tab.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _data = require("@wordpress/data");

var _blockTypesList = _interopRequireDefault(require("../block-types-list"));

var _useClipboardBlock = _interopRequireDefault(require("./hooks/use-clipboard-block"));

var _store = require("../../store");

var _useBlockTypeImpressions = _interopRequireDefault(require("./hooks/use-block-type-impressions"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const NON_BLOCK_CATEGORIES = ['reusable'];

function BlockTypesTab({
  onSelect,
  rootClientId,
  listProps
}) {
  const clipboardBlock = (0, _useClipboardBlock.default)(rootClientId);
  const {
    blockTypes
  } = (0, _data.useSelect)(select => {
    const {
      getInserterItems
    } = select(_store.store);
    const allItems = getInserterItems(rootClientId);
    const blockItems = allItems.filter(({
      category
    }) => !NON_BLOCK_CATEGORIES.includes(category));
    return {
      blockTypes: clipboardBlock ? [clipboardBlock, ...blockItems] : blockItems
    };
  }, [rootClientId]);
  const {
    items,
    trackBlockTypeSelected
  } = (0, _useBlockTypeImpressions.default)(blockTypes);

  const handleSelect = (...args) => {
    trackBlockTypeSelected(...args);
    onSelect(...args);
  };

  return (0, _element.createElement)(_blockTypesList.default, {
    name: "Blocks",
    items: items,
    onSelect: handleSelect,
    listProps: listProps
  });
}

var _default = BlockTypesTab;
exports.default = _default;
