// @wordpress/components/build/segmented-control/segmented-control-backdrop.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _styles = require("./styles");

/**
 * WordPress dependencies
 */
function SegmentedControlBackdrop({
  containerRef,
  containerWidth,
  state
}) {
  const [left, setLeft] = (0, _element.useState)(0);
  const [width, setWidth] = (0, _element.useState)(0);
  const [canAnimate, setCanAnimate] = (0, _element.useState)(false);
  (0, _element.useEffect)(() => {
    const containerNode = containerRef === null || containerRef === void 0 ? void 0 : containerRef.current;
    if (!containerNode) return;
    /**
     * Workaround for Reakit
     */

    const targetNode = containerNode.querySelector(`[data-value="${state}"]`);
    if (!targetNode) return;
    const {
      x: parentX
    } = containerNode.getBoundingClientRect();
    const {
      width: offsetWidth,
      x
    } = targetNode.getBoundingClientRect();
    const borderWidth = 1;
    const offsetLeft = x - parentX - borderWidth;
    setLeft(offsetLeft);
    setWidth(offsetWidth);

    if (!canAnimate) {
      window.requestAnimationFrame(() => {
        setCanAnimate(true);
      });
    }
  }, [canAnimate, containerRef, containerWidth, state]);
  return (0, _element.createElement)(_styles.BackdropView, {
    role: "presentation",
    style: {
      transform: `translateX(${left}px)`,
      transition: canAnimate ? undefined : 'none',
      width
    }
  });
}

var _default = (0, _element.memo)(SegmentedControlBackdrop);

exports.default = _default;
