// @wordpress/components/build-module/ui/color-picker/component.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
// eslint-disable-next-line no-restricted-imports

/**
 * WordPress dependencies
 */
import { useState } from '@wordpress/element';
import { moreVertical } from '@wordpress/icons';
import { useDebounce } from '@wordpress/compose';
import { __ } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import { useContextSystem, contextConnect } from '../context';
import { HStack } from '../../h-stack';
import Button from '../../button';
import { Spacer } from '../../spacer';
import { ColorfulWrapper, SelectControl } from './styles';
import { ColorDisplay } from './color-display';
import { ColorInput } from './color-input';
import { Picker } from './picker';
import { useControlledValue } from '../../utils/hooks';
const options = [{
  label: 'RGB',
  value: 'rgb'
}, {
  label: 'HSL',
  value: 'hsl'
}, {
  label: 'Hex',
  value: 'hex'
}];

const getSafeColor = color => {
  return color ? {
    a: 1,
    ...color
  } : {
    h: 0,
    s: 0,
    l: 100,
    a: 1
  };
};

const ColorPicker = (props, forwardedRef) => {
  const {
    enableAlpha = false,
    color: colorProp,
    onChange,
    defaultValue,
    copyFormat
  } = useContextSystem(props, 'ColorPicker');
  const [color, setColor] = useControlledValue({
    onChange,
    value: colorProp,
    defaultValue
  }); // Debounce to prevent rapid changes from conflicting with one another.

  const debouncedSetColor = useDebounce(setColor);

  const handleChange = nextValue => {
    debouncedSetColor(nextValue);
  }; // Use a safe default value for the color and remove the possibility of `undefined`.


  const safeColor = getSafeColor(color);
  const [showInputs, setShowInputs] = useState(false);
  const [colorType, setColorType] = useState(copyFormat || 'hex');
  return createElement(ColorfulWrapper, {
    ref: forwardedRef
  }, createElement(Picker, {
    onChange: handleChange,
    color: safeColor,
    enableAlpha: enableAlpha
  }), createElement(HStack, {
    justify: "space-between"
  }, showInputs ? createElement(SelectControl, {
    options: options,
    value: colorType,
    onChange: nextColorType => setColorType(nextColorType),
    label: __('Color format'),
    hideLabelFromVision: true
  }) : createElement(ColorDisplay, {
    color: safeColor,
    colorType: copyFormat || colorType,
    enableAlpha: enableAlpha
  }), createElement(Button, {
    onClick: () => setShowInputs(!showInputs),
    icon: moreVertical,
    isPressed: showInputs,
    label: showInputs ? __('Hide detailed inputs') : __('Show detailed inputs')
  })), createElement(Spacer, null), showInputs && createElement(ColorInput, {
    colorType: colorType,
    color: safeColor,
    onChange: handleChange,
    enableAlpha: enableAlpha
  }));
};

const ConnectedColorPicker = contextConnect(ColorPicker, 'ColorPicker');
export default ConnectedColorPicker;
