// @wordpress/components/build-module/ui/color-picker/hsl-input.js
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
import { InputWithSlider } from './input-with-slider';
export const HslInput = ({
  color,
  onChange,
  enableAlpha
}) => {
  const {
    h,
    s,
    l,
    a
  } = color;
  return createElement(Fragment, null, createElement(InputWithSlider, {
    min: 1,
    max: 360,
    label: "Hue",
    abbreviation: "H",
    value: Math.trunc(h),
    onChange: nextH => onChange({
      h: nextH,
      s,
      l,
      a
    })
  }), createElement(InputWithSlider, {
    min: 0,
    max: 100,
    label: "Saturation",
    abbreviation: "S",
    value: s <= 1 ? Math.trunc(100 * s) : s,
    onChange: nextS => onChange({
      h,
      s: nextS / 100,
      l,
      a
    })
  }), createElement(InputWithSlider, {
    min: 0,
    max: 100,
    label: "Lightness",
    abbreviation: "L",
    value: l <= 1 ? Math.trunc(100 * l) : l,
    onChange: nextL => onChange({
      h,
      s,
      l: nextL / 100,
      a
    })
  }), enableAlpha && createElement(InputWithSlider, {
    min: 0,
    max: 100,
    label: "Alpha",
    abbreviation: "A",
    value: Math.trunc(100 * a),
    onChange: nextA => onChange({
      h,
      s,
      l,
      a: nextA / 100
    })
  }));
};
