// @wordpress/components/build-module/ui/color-picker/input-with-slider.js
import { createElement } from "@wordpress/element";

/**
 * Internal dependencies
 */
import NumberControl from '../../number-control';
import { HStack } from '../../h-stack';
import { Text } from '../../text';
import { Spacer } from '../../spacer';
import { space } from '../utils/space';
import { RangeControl } from './styles';
export const InputWithSlider = ({
  min,
  max,
  label,
  abbreviation,
  onChange,
  value
}) => {
  return createElement(Spacer, {
    as: HStack
  }, createElement(NumberControl, {
    __unstableInputWidth: "5em",
    min: min,
    max: max,
    label: label,
    hideLabelFromVision: true,
    value: value,
    onChange: onChange,
    prefix: createElement(Spacer, {
      as: Text,
      paddingLeft: space(1),
      color: "blue"
    }, abbreviation),
    hideHTMLArrows: true
  }), createElement(RangeControl, {
    label: label,
    hideLabelFromVision: true,
    min: min,
    max: max,
    value: value,
    onChange: onChange,
    withInputField: false
  }));
};
