// @wordpress/components/build-module/ui/color-picker/rgb-input.js
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import colorize from 'tinycolor2';
/**
 * Internal dependencies
 */

import { InputWithSlider } from './input-with-slider';
export const RgbInput = ({
  color,
  onChange,
  enableAlpha
}) => {
  const {
    r,
    g,
    b,
    a
  } = colorize(color).toRgb();
  return createElement(Fragment, null, createElement(InputWithSlider, {
    min: 0,
    max: 255,
    label: "Red",
    abbreviation: "R",
    value: r,
    onChange: nextR => onChange(colorize({
      r: nextR,
      g,
      b,
      a
    }).toHsl())
  }), createElement(InputWithSlider, {
    min: 0,
    max: 255,
    label: "Green",
    abbreviation: "G",
    value: g,
    onChange: nextG => onChange(colorize({
      r,
      g: nextG,
      b,
      a
    }).toHsl())
  }), createElement(InputWithSlider, {
    min: 0,
    max: 255,
    label: "Blue",
    abbreviation: "B",
    value: b,
    onChange: nextB => onChange(colorize({
      r,
      g,
      b: nextB,
      a
    }).toHsl())
  }), enableAlpha && createElement(InputWithSlider, {
    min: 0,
    max: 100,
    label: "Alpha",
    abbreviation: "A",
    value: Math.trunc(a * 100),
    onChange: nextA => onChange(colorize({
      r,
      g,
      b,
      a: nextA / 100
    }).toHsl())
  }));
};
