// @wordpress/components/build/mobile/link-picker/link-picker-results.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LinkPickerResults;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _reactNative = require("react-native");

var _lodash = require("lodash");

var _components = require("@wordpress/components");

var _data = require("@wordpress/data");

var _styles = _interopRequireDefault(require("./styles.scss"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const PER_PAGE = 20;
const REQUEST_DEBOUNCE_DELAY = 400;
const MINIMUM_QUERY_SIZE = 2;

const meetsThreshold = query => MINIMUM_QUERY_SIZE <= query.length;

function LinkPickerResults({
  query,
  onLinkPicked,
  directEntry
}) {
  const [links, setLinks] = (0, _element.useState)([directEntry]);
  const [hasAllSuggestions, setHasAllSuggestions] = (0, _element.useState)(false);
  const nextPage = (0, _element.useRef)(1);
  const pendingRequest = (0, _element.useRef)();

  const clearRequest = () => {
    pendingRequest.current = null;
  }; // a stable debounced function to fetch suggestions and append


  const {
    fetchMoreSuggestions
  } = (0, _data.useSelect)(select => {
    const {
      getSettings
    } = select('core/block-editor');

    const fetchLinkSuggestions = async ({
      search
    }) => {
      if (meetsThreshold(search)) {
        return await getSettings().__experimentalFetchLinkSuggestions(search, {
          page: nextPage.current,
          type: 'post',
          perPage: PER_PAGE
        });
      }
    };

    const fetchMore = async ({
      query: search,
      links: currentSuggestions
    }) => {
      // return early if we've already detected the end of data or we are
      // already awaiting a response
      if (hasAllSuggestions || pendingRequest.current) {
        return;
      }

      const request = fetchLinkSuggestions({
        search
      });
      pendingRequest.current = request;
      const suggestions = await request; // only update links for the most recent request

      if (suggestions && request === pendingRequest.current) {
        // since we don't have the response header, we check if the results
        // are truncated to determine we've reached the end
        if (suggestions.length < PER_PAGE) {
          setHasAllSuggestions(true);
        }

        setLinks([...currentSuggestions, ...suggestions]);
        nextPage.current++;
      }

      clearRequest();
    };

    return {
      fetchMoreSuggestions: (0, _lodash.debounce)(fetchMore, REQUEST_DEBOUNCE_DELAY)
    };
  }, []); // prevent setting state when unmounted

  (0, _element.useEffect)(() => clearRequest, []); // any time the query changes, we reset pagination

  (0, _element.useEffect)(() => {
    clearRequest();
    nextPage.current = 1;
    setHasAllSuggestions(false);
    setLinks([directEntry]);
    fetchMoreSuggestions({
      query,
      links: [directEntry]
    });
  }, [query]);

  const onEndReached = () => fetchMoreSuggestions({
    query,
    links
  });

  const spinner = !hasAllSuggestions && meetsThreshold(query) && (0, _element.createElement)(_reactNative.View, {
    style: _styles.default.spinner
  }, (0, _element.createElement)(_reactNative.ActivityIndicator, {
    animating: true
  }));
  return (0, _element.createElement)(_components.BottomSheetConsumer, null, ({
    listProps
  }) => (0, _element.createElement)(_reactNative.FlatList, (0, _extends2.default)({
    data: links,
    keyboardShouldPersistTaps: "always",
    renderItem: ({
      item
    }) => (0, _element.createElement)(_components.BottomSheet.LinkSuggestionItemCell, {
      suggestion: item,
      onLinkPicked: onLinkPicked
    }),
    keyExtractor: ({
      url,
      type
    }) => `${url}-${type}`,
    onEndReached: onEndReached,
    onEndReachedThreshold: 0.1,
    initialNumToRender: PER_PAGE,
    ListFooterComponent: spinner
  }, listProps, {
    contentContainerStyle: [...listProps.contentContainerStyle, _styles.default.list]
  })));
}
