// @wordpress/components/build/ui/context/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  ContextSystemProvider: true,
  useComponentsContext: true,
  contextConnect: true,
  hasConnectNamespace: true,
  getConnectNamespace: true,
  useContextSystem: true
};
Object.defineProperty(exports, "ContextSystemProvider", {
  enumerable: true,
  get: function () {
    return _contextSystemProvider.ContextSystemProvider;
  }
});
Object.defineProperty(exports, "useComponentsContext", {
  enumerable: true,
  get: function () {
    return _contextSystemProvider.useComponentsContext;
  }
});
Object.defineProperty(exports, "contextConnect", {
  enumerable: true,
  get: function () {
    return _contextConnect.contextConnect;
  }
});
Object.defineProperty(exports, "hasConnectNamespace", {
  enumerable: true,
  get: function () {
    return _contextConnect.hasConnectNamespace;
  }
});
Object.defineProperty(exports, "getConnectNamespace", {
  enumerable: true,
  get: function () {
    return _contextConnect.getConnectNamespace;
  }
});
Object.defineProperty(exports, "useContextSystem", {
  enumerable: true,
  get: function () {
    return _useContextSystem.useContextSystem;
  }
});

var _contextSystemProvider = require("./context-system-provider");

var _contextConnect = require("./context-connect");

var _useContextSystem = require("./use-context-system");

var _polymorphicComponent = require("./polymorphic-component");

Object.keys(_polymorphicComponent).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _polymorphicComponent[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _polymorphicComponent[key];
    }
  });
});
