// @wordpress/block-editor/build/components/inserter-draggable-blocks/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _draggableChip = _interopRequireDefault(require("../block-draggable/draggable-chip"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const InserterDraggableBlocks = ({
  isEnabled,
  blocks,
  icon,
  children
}) => {
  const transferData = {
    type: 'inserter',
    blocks
  };
  return (0, _element.createElement)(_components.Draggable, {
    __experimentalTransferDataType: "wp-blocks",
    transferData: transferData,
    __experimentalDragComponent: (0, _element.createElement)(_draggableChip.default, {
      count: blocks.length,
      icon: icon
    })
  }, ({
    onDraggableStart,
    onDraggableEnd
  }) => {
    return children({
      draggable: isEnabled,
      onDragStart: isEnabled ? onDraggableStart : undefined,
      onDragEnd: isEnabled ? onDraggableEnd : undefined
    });
  });
};

var _default = InserterDraggableBlocks;
exports.default = _default;
