// @wordpress/block-editor/build/components/url-input/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = URLInput;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _reactNative = require("react-native");

var _i18n = require("@wordpress/i18n");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
function URLInput({
  value = '',
  autoFocus = true,
  onChange,
  ...extraProps
}) {
  /* eslint-disable jsx-a11y/no-autofocus */
  return (0, _element.createElement)(_reactNative.TextInput, (0, _extends2.default)({
    autoFocus: autoFocus,
    editable: true,
    selectTextOnFocus: true,
    autoCapitalize: "none",
    autoCorrect: false,
    textContentType: "URL",
    value: value,
    onChangeText: onChange,
    placeholder: (0, _i18n.__)('Paste URL')
  }, extraProps));
  /* eslint-enable jsx-a11y/no-autofocus */
}
