// @wordpress/components/build/flyout/flyout/component.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _reakit = require("reakit");

var _context = require("../../ui/context");

var _context2 = require("../context");

var _utils = require("../utils");

var _flyoutContent = _interopRequireDefault(require("../flyout-content"));

var _hooks = require("../../utils/hooks");

var _hook = require("./hook");

/**
 * External dependencies
 */
// eslint-disable-next-line no-restricted-imports

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 *
 * @param {import('../../ui/context').PolymorphicComponentProps<import('../types').Props, 'div', false>} props
 * @param {import('react').Ref<any>}                                                                     forwardedRef
 */
function Flyout(props, forwardedRef) {
  const {
    children,
    elevation,
    label,
    maxWidth,
    onVisibleChange,
    trigger,
    flyoutState,
    ...otherProps
  } = (0, _hook.useFlyout)(props);
  const resizeListener = (0, _utils.useFlyoutResizeUpdater)({
    onResize: flyoutState.unstable_update
  });
  const uniqueId = `flyout-${flyoutState.baseId}`;
  const labelId = label || uniqueId;
  const contextProps = (0, _element.useMemo)(() => ({
    label: labelId,
    flyoutState
  }), [labelId, flyoutState]);
  const triggerContent = (0, _element.useCallback)(triggerProps => {
    return (0, _element.cloneElement)(trigger, triggerProps);
  }, [trigger]);
  (0, _hooks.useUpdateEffect)(() => {
    onVisibleChange === null || onVisibleChange === void 0 ? void 0 : onVisibleChange(flyoutState.visible);
  }, [flyoutState.visible]);
  return (0, _element.createElement)(_context2.FlyoutContext.Provider, {
    value: contextProps
  }, trigger && (0, _element.createElement)(_reakit.PopoverDisclosure, (0, _extends2.default)({}, flyoutState, {
    ref: trigger.ref
  }, trigger.props), triggerContent), (0, _element.createElement)(_reakit.Portal, null, (0, _element.createElement)(_flyoutContent.default, (0, _extends2.default)({
    ref: forwardedRef
  }, otherProps, {
    elevation: elevation,
    maxWidth: maxWidth
  }), resizeListener, children)));
}
/**
 * `Flyout` is a component to render a floating content modal.
 * It is similar in purpose to a tooltip, but renders content of any sort,
 * not only simple text.
 *
 * @example
 * ```jsx
 * import { Button, __experimentalFlyout as Flyout, __experimentalText as } from '@wordpress/components';
 *
 * function Example() {
 * 	return (
 * 		<Flyout trigger={ <Button>Show/Hide content</Button> }>
 *			<Text>Code is Poetry</Text>
 * 		</Flyout>
 * 	);
 * }
 * ```
 */


const ConnectedFlyout = (0, _context.contextConnect)(Flyout, 'Flyout');
var _default = ConnectedFlyout;
exports.default = _default;
