// @wordpress/block-editor/build/components/list-view/leaf.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ListViewLeaf;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _web = require("react-spring/web.cjs");

var _classnames = _interopRequireDefault(require("classnames"));

var _components = require("@wordpress/components");

var _useMovingAnimation = _interopRequireDefault(require("../use-moving-animation"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const AnimatedTreeGridRow = (0, _web.animated)(_components.__experimentalTreeGridRow);

function ListViewLeaf({
  isSelected,
  position,
  level,
  rowCount,
  children,
  className,
  path,
  ...props
}) {
  const ref = (0, _useMovingAnimation.default)({
    isSelected,
    adjustScrolling: false,
    enableAnimation: true,
    triggerAnimationOnChange: path.join('_')
  });
  return (0, _element.createElement)(AnimatedTreeGridRow, (0, _extends2.default)({
    ref: ref,
    className: (0, _classnames.default)('block-editor-list-view-leaf', className),
    level: level,
    positionInSet: position,
    setSize: rowCount
  }, props), children);
}
