// @wordpress/block-editor/build/components/letter-spacing-control/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LetterSpacingControl;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _useSetting = _interopRequireDefault(require("../../components/use-setting"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * Control for letter-spacing.
 *
 * @param {Object}   props          Component props.
 * @param {string}   props.value    Currently selected letter-spacing.
 * @param {Function} props.onChange Handles change in letter-spacing selection.
 * @return {WPElement}                      Letter-spacing control.
 */
function LetterSpacingControl({
  value,
  onChange
}) {
  const units = (0, _components.__experimentalUseCustomUnits)({
    availableUnits: (0, _useSetting.default)('spacing.units') || ['px', 'em', 'rem'],
    defaultValues: {
      px: '2',
      em: '.2',
      rem: '.2'
    }
  });
  return (0, _element.createElement)(_components.__experimentalUnitControl, {
    label: (0, _i18n.__)('Letter-spacing'),
    value: value,
    __unstableInputWidth: "60px",
    units: units,
    onChange: onChange
  });
}
