// @wordpress/components/build/button/deprecated.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _deprecated = _interopRequireDefault(require("@wordpress/deprecated"));

var _button = _interopRequireDefault(require("../button"));

// @ts-nocheck

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function IconButton({
  labelPosition,
  size,
  tooltip,
  label,
  ...props
}, ref) {
  (0, _deprecated.default)('wp.components.IconButton', {
    since: '5.4',
    alternative: 'wp.components.Button'
  });
  return (0, _element.createElement)(_button.default, (0, _extends2.default)({}, props, {
    ref: ref,
    tooltipPosition: labelPosition,
    iconSize: size,
    showTooltip: tooltip !== undefined ? !!tooltip : undefined,
    label: tooltip || label
  }));
}

var _default = (0, _element.forwardRef)(IconButton);

exports.default = _default;
