// @wordpress/block-editor/build/components/url-popover/link-editor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LinkEditor;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

var _icons = require("@wordpress/icons");

var _urlInput = _interopRequireDefault(require("../url-input"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function LinkEditor({
  autocompleteRef,
  className,
  onChangeInputValue,
  value,
  ...props
}) {
  return (0, _element.createElement)("form", (0, _extends2.default)({
    className: (0, _classnames.default)('block-editor-url-popover__link-editor', className)
  }, props), (0, _element.createElement)(_urlInput.default, {
    value: value,
    onChange: onChangeInputValue,
    autocompleteRef: autocompleteRef
  }), (0, _element.createElement)(_components.Button, {
    icon: _icons.keyboardReturn,
    label: (0, _i18n.__)('Apply'),
    type: "submit"
  }));
}
