// @wordpress/components/build/resizable-box/resize-tooltip/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _lodash = require("lodash");

var _classnames = _interopRequireDefault(require("classnames"));

var _label = _interopRequireDefault(require("./label"));

var _utils = require("./utils");

var _resizeTooltip = require("./styles/resize-tooltip.styles");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function ResizeTooltip({
  axis,
  className,
  fadeTimeout = 180,
  isVisible = true,
  labelRef,
  onResize = _lodash.noop,
  position = _utils.POSITIONS.bottom,
  showPx = true,
  zIndex = 1000,
  ...props
}, ref) {
  const {
    label,
    resizeListener
  } = (0, _utils.useResizeLabel)({
    axis,
    fadeTimeout,
    onResize,
    showPx,
    position
  });
  if (!isVisible) return null;
  const classes = (0, _classnames.default)('components-resize-tooltip', className);
  return (0, _element.createElement)(_resizeTooltip.Root, (0, _extends2.default)({
    "aria-hidden": "true",
    className: classes,
    ref: ref
  }, props), resizeListener, (0, _element.createElement)(_label.default, {
    "aria-hidden": props['aria-hidden'],
    fadeTimeout: fadeTimeout,
    isVisible: isVisible,
    label: label,
    position: position,
    ref: labelRef,
    zIndex: zIndex
  }));
}

const ForwardedComponent = (0, _element.forwardRef)(ResizeTooltip);
var _default = ForwardedComponent;
exports.default = _default;
