// @wordpress/components/build/ui/color-picker/hsl-input.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HslInput = void 0;

var _element = require("@wordpress/element");

var _inputWithSlider = require("./input-with-slider");

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
const HslInput = ({
  color,
  onChange,
  enableAlpha
}) => {
  const {
    h,
    s,
    l,
    a
  } = color;
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_inputWithSlider.InputWithSlider, {
    min: 1,
    max: 360,
    label: "Hue",
    abbreviation: "H",
    value: Math.trunc(h),
    onChange: nextH => onChange({
      h: nextH,
      s,
      l,
      a
    })
  }), (0, _element.createElement)(_inputWithSlider.InputWithSlider, {
    min: 0,
    max: 100,
    label: "Saturation",
    abbreviation: "S",
    value: s <= 1 ? Math.trunc(100 * s) : s,
    onChange: nextS => onChange({
      h,
      s: nextS / 100,
      l,
      a
    })
  }), (0, _element.createElement)(_inputWithSlider.InputWithSlider, {
    min: 0,
    max: 100,
    label: "Lightness",
    abbreviation: "L",
    value: l <= 1 ? Math.trunc(100 * l) : l,
    onChange: nextL => onChange({
      h,
      s,
      l: nextL / 100,
      a
    })
  }), enableAlpha && (0, _element.createElement)(_inputWithSlider.InputWithSlider, {
    min: 0,
    max: 100,
    label: "Alpha",
    abbreviation: "A",
    value: Math.trunc(100 * a),
    onChange: nextA => onChange({
      h,
      s,
      l,
      a: nextA / 100
    })
  }));
};

exports.HslInput = HslInput;
