// @wordpress/components/build/ui/color-picker/input-with-slider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InputWithSlider = void 0;

var _element = require("@wordpress/element");

var _numberControl = _interopRequireDefault(require("../../number-control"));

var _hStack = require("../../h-stack");

var _text = require("../../text");

var _spacer = require("../../spacer");

var _space = require("../utils/space");

var _styles = require("./styles");

/**
 * Internal dependencies
 */
const InputWithSlider = ({
  min,
  max,
  label,
  abbreviation,
  onChange,
  value
}) => {
  return (0, _element.createElement)(_spacer.Spacer, {
    as: _hStack.HStack
  }, (0, _element.createElement)(_numberControl.default, {
    __unstableInputWidth: "5em",
    min: min,
    max: max,
    label: label,
    hideLabelFromVision: true,
    value: value,
    onChange: onChange,
    prefix: (0, _element.createElement)(_spacer.Spacer, {
      as: _text.Text,
      paddingLeft: (0, _space.space)(1),
      color: "blue"
    }, abbreviation),
    hideHTMLArrows: true
  }), (0, _element.createElement)(_styles.RangeControl, {
    label: label,
    hideLabelFromVision: true,
    min: min,
    max: max,
    value: value,
    onChange: onChange,
    withInputField: false
  }));
};

exports.InputWithSlider = InputWithSlider;
