// @wordpress/components/build/ui/color-picker/rgb-input.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RgbInput = void 0;

var _element = require("@wordpress/element");

var _tinycolor = _interopRequireDefault(require("tinycolor2"));

var _inputWithSlider = require("./input-with-slider");

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
const RgbInput = ({
  color,
  onChange,
  enableAlpha
}) => {
  const {
    r,
    g,
    b,
    a
  } = (0, _tinycolor.default)(color).toRgb();
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_inputWithSlider.InputWithSlider, {
    min: 0,
    max: 255,
    label: "Red",
    abbreviation: "R",
    value: r,
    onChange: nextR => onChange((0, _tinycolor.default)({
      r: nextR,
      g,
      b,
      a
    }).toHsl())
  }), (0, _element.createElement)(_inputWithSlider.InputWithSlider, {
    min: 0,
    max: 255,
    label: "Green",
    abbreviation: "G",
    value: g,
    onChange: nextG => onChange((0, _tinycolor.default)({
      r,
      g: nextG,
      b,
      a
    }).toHsl())
  }), (0, _element.createElement)(_inputWithSlider.InputWithSlider, {
    min: 0,
    max: 255,
    label: "Blue",
    abbreviation: "B",
    value: b,
    onChange: nextB => onChange((0, _tinycolor.default)({
      r,
      g,
      b: nextB,
      a
    }).toHsl())
  }), enableAlpha && (0, _element.createElement)(_inputWithSlider.InputWithSlider, {
    min: 0,
    max: 100,
    label: "Alpha",
    abbreviation: "A",
    value: Math.trunc(a * 100),
    onChange: nextA => onChange((0, _tinycolor.default)({
      r,
      g,
      b,
      a: nextA / 100
    }).toHsl())
  }));
};

exports.RgbInput = RgbInput;
