// @wordpress/components/build/form-toggle/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classnames = _interopRequireDefault(require("classnames"));

var _lodash = require("lodash");

/**
 * External dependencies
 */
function FormToggle({
  className,
  checked,
  id,
  disabled,
  onChange = _lodash.noop,
  ...props
}) {
  const wrapperClasses = (0, _classnames.default)('components-form-toggle', className, {
    'is-checked': checked,
    'is-disabled': disabled
  });
  return (0, _element.createElement)("span", {
    className: wrapperClasses
  }, (0, _element.createElement)("input", (0, _extends2.default)({
    className: "components-form-toggle__input",
    id: id,
    type: "checkbox",
    checked: checked,
    onChange: onChange,
    disabled: disabled
  }, props)), (0, _element.createElement)("span", {
    className: "components-form-toggle__track"
  }), (0, _element.createElement)("span", {
    className: "components-form-toggle__thumb"
  }));
}

var _default = FormToggle;
exports.default = _default;
