// @wordpress/block-editor/build-module/components/link-control/link-preview.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
import { Button, ExternalLink, __experimentalText as Text } from '@wordpress/components';
import { filterURLForDisplay, safeDecodeURI } from '@wordpress/url';
import { Icon, globe } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import { ViewerSlot } from './viewer-slot';
import useRichUrlData from './use-rich-url-data';
export default function LinkPreview({
  value,
  onEditClick,
  hasRichPreviews = false
}) {
  // Avoid fetching if rich previews are not desired.
  const showRichPreviews = hasRichPreviews ? value === null || value === void 0 ? void 0 : value.url : null;
  const {
    richData,
    isFetching
  } = useRichUrlData(showRichPreviews); // Rich data may be an empty object so test for that.

  const hasRichData = richData && Object.keys(richData).length;
  const displayURL = value && filterURLForDisplay(safeDecodeURI(value.url), 16) || '';
  return createElement("div", {
    "aria-label": __('Currently selected'),
    "aria-selected": "true",
    className: classnames('block-editor-link-control__search-item', {
      'is-current': true,
      'is-rich': hasRichData,
      'is-fetching': !!isFetching,
      'is-preview': true
    })
  }, createElement("div", {
    className: "block-editor-link-control__search-item-top"
  }, createElement("span", {
    className: "block-editor-link-control__search-item-header"
  }, createElement("span", {
    className: classnames('block-editor-link-control__search-item-icon', {
      'is-image': richData === null || richData === void 0 ? void 0 : richData.icon
    })
  }, richData !== null && richData !== void 0 && richData.icon ? createElement("img", {
    src: richData === null || richData === void 0 ? void 0 : richData.icon,
    alt: ""
  }) : createElement(Icon, {
    icon: globe
  })), createElement("span", {
    className: "block-editor-link-control__search-item-details"
  }, createElement(ExternalLink, {
    className: "block-editor-link-control__search-item-title",
    href: value.url
  }, (richData === null || richData === void 0 ? void 0 : richData.title) || (value === null || value === void 0 ? void 0 : value.title) || displayURL), (value === null || value === void 0 ? void 0 : value.url) && createElement("span", {
    className: "block-editor-link-control__search-item-info"
  }, displayURL))), createElement(Button, {
    variant: "secondary",
    onClick: () => onEditClick(),
    className: "block-editor-link-control__search-item-action"
  }, __('Edit')), createElement(ViewerSlot, {
    fillProps: value
  })), (hasRichData && ((richData === null || richData === void 0 ? void 0 : richData.image) || (richData === null || richData === void 0 ? void 0 : richData.description)) || isFetching) && createElement("div", {
    className: "block-editor-link-control__search-item-bottom"
  }, ((richData === null || richData === void 0 ? void 0 : richData.image) || isFetching) && createElement("div", {
    "aria-hidden": !(richData !== null && richData !== void 0 && richData.image),
    className: classnames('block-editor-link-control__search-item-image', {
      'is-placeholder': !(richData !== null && richData !== void 0 && richData.image)
    })
  }, (richData === null || richData === void 0 ? void 0 : richData.image) && createElement("img", {
    src: richData === null || richData === void 0 ? void 0 : richData.image,
    alt: ""
  })), ((richData === null || richData === void 0 ? void 0 : richData.description) || isFetching) && createElement("div", {
    "aria-hidden": !(richData !== null && richData !== void 0 && richData.description),
    className: classnames('block-editor-link-control__search-item-description', {
      'is-placeholder': !(richData !== null && richData !== void 0 && richData.description)
    })
  }, (richData === null || richData === void 0 ? void 0 : richData.description) && createElement(Text, {
    truncate: true,
    numberOfLines: "2"
  }, richData.description))));
}
