// @wordpress/block-editor/build/components/border-radius-control/linked-button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LinkedButton;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _components = require("@wordpress/components");

var _icons = require("@wordpress/icons");

var _i18n = require("@wordpress/i18n");

/**
 * WordPress dependencies
 */
function LinkedButton({
  isLinked,
  ...props
}) {
  const label = isLinked ? (0, _i18n.__)('Unlink Radii') : (0, _i18n.__)('Link Radii');
  return (0, _element.createElement)(_components.Tooltip, {
    text: label
  }, (0, _element.createElement)(_components.Button, (0, _extends2.default)({}, props, {
    className: "component-border-radius-control__linked-button",
    isPrimary: isLinked,
    isSecondary: !isLinked,
    isSmall: true,
    icon: isLinked ? _icons.link : _icons.linkOff,
    iconSize: 16,
    "aria-label": label
  })));
}
