// @wordpress/block-editor/build/components/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  BlockAlignmentControl: true,
  BlockAlignmentToolbar: true,
  BlockContextProvider: true,
  BlockControls: true,
  BlockFormatControls: true,
  BlockEdit: true,
  useBlockEditContext: true,
  BlockIcon: true,
  BlockVerticalAlignmentToolbar: true,
  BlockVerticalAlignmentControl: true,
  AlignmentControl: true,
  AlignmentToolbar: true,
  InnerBlocks: true,
  InspectorAdvancedControls: true,
  InspectorControls: true,
  JustifyToolbar: true,
  JustifyContentControl: true,
  LineHeightControl: true,
  PlainText: true,
  RichText: true,
  RichTextShortcut: true,
  RichTextToolbarButton: true,
  __unstableRichTextInputEvent: true,
  MediaPlaceholder: true,
  MediaUpload: true,
  MEDIA_TYPE_IMAGE: true,
  MEDIA_TYPE_VIDEO: true,
  MEDIA_TYPE_AUDIO: true,
  MEDIA_TYPE_ANY: true,
  MediaUploadProgress: true,
  BlockMediaUpdateProgress: true,
  URLInput: true,
  BlockInvalidWarning: true,
  BlockCaption: true,
  Caption: true,
  PanelColorSettings: true,
  __experimentalPanelColorGradientSettings: true,
  useSetting: true,
  __experimentalUseNoRecursiveRenders: true,
  Warning: true,
  BottomSheetSettings: true,
  BlockSettingsButton: true,
  blockSettingsScreens: true,
  VideoPlayer: true,
  VIDEO_ASPECT_RATIO: true,
  BlockList: true,
  BlockMover: true,
  BlockToolbar: true,
  BlockVariationPicker: true,
  BlockStyles: true,
  DefaultBlockAppender: true,
  __unstableEditorStyles: true,
  Inserter: true,
  useBlockProps: true,
  FloatingToolbar: true,
  BlockEditorProvider: true
};
Object.defineProperty(exports, "BlockAlignmentControl", {
  enumerable: true,
  get: function () {
    return _blockAlignmentControl.BlockAlignmentControl;
  }
});
Object.defineProperty(exports, "BlockAlignmentToolbar", {
  enumerable: true,
  get: function () {
    return _blockAlignmentControl.BlockAlignmentToolbar;
  }
});
Object.defineProperty(exports, "BlockContextProvider", {
  enumerable: true,
  get: function () {
    return _blockContext.BlockContextProvider;
  }
});
Object.defineProperty(exports, "BlockControls", {
  enumerable: true,
  get: function () {
    return _blockControls.default;
  }
});
Object.defineProperty(exports, "BlockFormatControls", {
  enumerable: true,
  get: function () {
    return _blockControls.BlockFormatControls;
  }
});
Object.defineProperty(exports, "BlockEdit", {
  enumerable: true,
  get: function () {
    return _blockEdit.default;
  }
});
Object.defineProperty(exports, "useBlockEditContext", {
  enumerable: true,
  get: function () {
    return _blockEdit.useBlockEditContext;
  }
});
Object.defineProperty(exports, "BlockIcon", {
  enumerable: true,
  get: function () {
    return _blockIcon.default;
  }
});
Object.defineProperty(exports, "BlockVerticalAlignmentToolbar", {
  enumerable: true,
  get: function () {
    return _blockVerticalAlignmentControl.BlockVerticalAlignmentToolbar;
  }
});
Object.defineProperty(exports, "BlockVerticalAlignmentControl", {
  enumerable: true,
  get: function () {
    return _blockVerticalAlignmentControl.BlockVerticalAlignmentControl;
  }
});
Object.defineProperty(exports, "AlignmentControl", {
  enumerable: true,
  get: function () {
    return _alignmentControl.AlignmentControl;
  }
});
Object.defineProperty(exports, "AlignmentToolbar", {
  enumerable: true,
  get: function () {
    return _alignmentControl.AlignmentToolbar;
  }
});
Object.defineProperty(exports, "InnerBlocks", {
  enumerable: true,
  get: function () {
    return _innerBlocks.default;
  }
});
Object.defineProperty(exports, "InspectorAdvancedControls", {
  enumerable: true,
  get: function () {
    return _inspectorAdvancedControls.default;
  }
});
Object.defineProperty(exports, "InspectorControls", {
  enumerable: true,
  get: function () {
    return _inspectorControls.default;
  }
});
Object.defineProperty(exports, "JustifyToolbar", {
  enumerable: true,
  get: function () {
    return _justifyContentControl.JustifyToolbar;
  }
});
Object.defineProperty(exports, "JustifyContentControl", {
  enumerable: true,
  get: function () {
    return _justifyContentControl.JustifyContentControl;
  }
});
Object.defineProperty(exports, "LineHeightControl", {
  enumerable: true,
  get: function () {
    return _lineHeightControl.default;
  }
});
Object.defineProperty(exports, "PlainText", {
  enumerable: true,
  get: function () {
    return _plainText.default;
  }
});
Object.defineProperty(exports, "RichText", {
  enumerable: true,
  get: function () {
    return _richText.default;
  }
});
Object.defineProperty(exports, "RichTextShortcut", {
  enumerable: true,
  get: function () {
    return _richText.RichTextShortcut;
  }
});
Object.defineProperty(exports, "RichTextToolbarButton", {
  enumerable: true,
  get: function () {
    return _richText.RichTextToolbarButton;
  }
});
Object.defineProperty(exports, "__unstableRichTextInputEvent", {
  enumerable: true,
  get: function () {
    return _richText.__unstableRichTextInputEvent;
  }
});
Object.defineProperty(exports, "MediaPlaceholder", {
  enumerable: true,
  get: function () {
    return _mediaPlaceholder.default;
  }
});
Object.defineProperty(exports, "MediaUpload", {
  enumerable: true,
  get: function () {
    return _mediaUpload.default;
  }
});
Object.defineProperty(exports, "MEDIA_TYPE_IMAGE", {
  enumerable: true,
  get: function () {
    return _mediaUpload.MEDIA_TYPE_IMAGE;
  }
});
Object.defineProperty(exports, "MEDIA_TYPE_VIDEO", {
  enumerable: true,
  get: function () {
    return _mediaUpload.MEDIA_TYPE_VIDEO;
  }
});
Object.defineProperty(exports, "MEDIA_TYPE_AUDIO", {
  enumerable: true,
  get: function () {
    return _mediaUpload.MEDIA_TYPE_AUDIO;
  }
});
Object.defineProperty(exports, "MEDIA_TYPE_ANY", {
  enumerable: true,
  get: function () {
    return _mediaUpload.MEDIA_TYPE_ANY;
  }
});
Object.defineProperty(exports, "MediaUploadProgress", {
  enumerable: true,
  get: function () {
    return _mediaUploadProgress.default;
  }
});
Object.defineProperty(exports, "BlockMediaUpdateProgress", {
  enumerable: true,
  get: function () {
    return _blockMediaUpdateProgress.default;
  }
});
Object.defineProperty(exports, "URLInput", {
  enumerable: true,
  get: function () {
    return _urlInput.default;
  }
});
Object.defineProperty(exports, "BlockInvalidWarning", {
  enumerable: true,
  get: function () {
    return _blockInvalidWarning.default;
  }
});
Object.defineProperty(exports, "BlockCaption", {
  enumerable: true,
  get: function () {
    return _blockCaption.default;
  }
});
Object.defineProperty(exports, "Caption", {
  enumerable: true,
  get: function () {
    return _caption.default;
  }
});
Object.defineProperty(exports, "PanelColorSettings", {
  enumerable: true,
  get: function () {
    return _panelColorSettings.default;
  }
});
Object.defineProperty(exports, "__experimentalPanelColorGradientSettings", {
  enumerable: true,
  get: function () {
    return _panelColorGradientSettings.default;
  }
});
Object.defineProperty(exports, "useSetting", {
  enumerable: true,
  get: function () {
    return _useSetting.default;
  }
});
Object.defineProperty(exports, "__experimentalUseNoRecursiveRenders", {
  enumerable: true,
  get: function () {
    return _useNoRecursiveRenders.default;
  }
});
Object.defineProperty(exports, "Warning", {
  enumerable: true,
  get: function () {
    return _warning.default;
  }
});
Object.defineProperty(exports, "BottomSheetSettings", {
  enumerable: true,
  get: function () {
    return _blockSettings.BottomSheetSettings;
  }
});
Object.defineProperty(exports, "BlockSettingsButton", {
  enumerable: true,
  get: function () {
    return _blockSettings.BlockSettingsButton;
  }
});
Object.defineProperty(exports, "blockSettingsScreens", {
  enumerable: true,
  get: function () {
    return _blockSettings.blockSettingsScreens;
  }
});
Object.defineProperty(exports, "VideoPlayer", {
  enumerable: true,
  get: function () {
    return _videoPlayer.default;
  }
});
Object.defineProperty(exports, "VIDEO_ASPECT_RATIO", {
  enumerable: true,
  get: function () {
    return _videoPlayer.VIDEO_ASPECT_RATIO;
  }
});
Object.defineProperty(exports, "BlockList", {
  enumerable: true,
  get: function () {
    return _blockList.default;
  }
});
Object.defineProperty(exports, "BlockMover", {
  enumerable: true,
  get: function () {
    return _blockMover.default;
  }
});
Object.defineProperty(exports, "BlockToolbar", {
  enumerable: true,
  get: function () {
    return _blockToolbar.default;
  }
});
Object.defineProperty(exports, "BlockVariationPicker", {
  enumerable: true,
  get: function () {
    return _blockVariationPicker.default;
  }
});
Object.defineProperty(exports, "BlockStyles", {
  enumerable: true,
  get: function () {
    return _blockStyles.default;
  }
});
Object.defineProperty(exports, "DefaultBlockAppender", {
  enumerable: true,
  get: function () {
    return _defaultBlockAppender.default;
  }
});
Object.defineProperty(exports, "__unstableEditorStyles", {
  enumerable: true,
  get: function () {
    return _editorStyles.default;
  }
});
Object.defineProperty(exports, "Inserter", {
  enumerable: true,
  get: function () {
    return _inserter.default;
  }
});
Object.defineProperty(exports, "useBlockProps", {
  enumerable: true,
  get: function () {
    return _useBlockProps.useBlockProps;
  }
});
Object.defineProperty(exports, "FloatingToolbar", {
  enumerable: true,
  get: function () {
    return _floatingToolbar.default;
  }
});
Object.defineProperty(exports, "BlockEditorProvider", {
  enumerable: true,
  get: function () {
    return _provider.default;
  }
});

var _blockAlignmentControl = require("./block-alignment-control");

var _blockContext = require("./block-context");

var _blockControls = _interopRequireWildcard(require("./block-controls"));

var _blockEdit = _interopRequireWildcard(require("./block-edit"));

var _blockIcon = _interopRequireDefault(require("./block-icon"));

var _blockVerticalAlignmentControl = require("./block-vertical-alignment-control");

var _colors = require("./colors");

Object.keys(_colors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _colors[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _colors[key];
    }
  });
});

var _gradients = require("./gradients");

Object.keys(_gradients).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _gradients[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _gradients[key];
    }
  });
});

var _fontSizes = require("./font-sizes");

Object.keys(_fontSizes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _fontSizes[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _fontSizes[key];
    }
  });
});

var _alignmentControl = require("./alignment-control");

var _innerBlocks = _interopRequireDefault(require("./inner-blocks"));

var _inspectorAdvancedControls = _interopRequireDefault(require("./inspector-advanced-controls"));

var _inspectorControls = _interopRequireDefault(require("./inspector-controls"));

var _justifyContentControl = require("./justify-content-control");

var _lineHeightControl = _interopRequireDefault(require("./line-height-control"));

var _plainText = _interopRequireDefault(require("./plain-text"));

var _richText = _interopRequireWildcard(require("./rich-text"));

var _mediaPlaceholder = _interopRequireDefault(require("./media-placeholder"));

var _mediaUpload = _interopRequireWildcard(require("./media-upload"));

var _mediaUploadProgress = _interopRequireDefault(require("./media-upload-progress"));

var _blockMediaUpdateProgress = _interopRequireDefault(require("./block-media-update-progress"));

var _urlInput = _interopRequireDefault(require("./url-input"));

var _blockInvalidWarning = _interopRequireDefault(require("./block-list/block-invalid-warning"));

var _blockCaption = _interopRequireDefault(require("./block-caption"));

var _caption = _interopRequireDefault(require("./caption"));

var _panelColorSettings = _interopRequireDefault(require("./panel-color-settings"));

var _panelColorGradientSettings = _interopRequireDefault(require("./colors-gradients/panel-color-gradient-settings"));

var _useSetting = _interopRequireDefault(require("./use-setting"));

var _useNoRecursiveRenders = _interopRequireDefault(require("./use-no-recursive-renders"));

var _warning = _interopRequireDefault(require("./warning"));

var _blockSettings = require("./block-settings");

var _videoPlayer = _interopRequireWildcard(require("./video-player"));

var _blockList = _interopRequireDefault(require("./block-list"));

var _blockMover = _interopRequireDefault(require("./block-mover"));

var _blockToolbar = _interopRequireDefault(require("./block-toolbar"));

var _blockVariationPicker = _interopRequireDefault(require("./block-variation-picker"));

var _blockStyles = _interopRequireDefault(require("./block-styles"));

var _defaultBlockAppender = _interopRequireDefault(require("./default-block-appender"));

var _editorStyles = _interopRequireDefault(require("./editor-styles"));

var _inserter = _interopRequireDefault(require("./inserter"));

var _useBlockProps = require("./block-list/use-block-props");

var _floatingToolbar = _interopRequireDefault(require("./floating-toolbar"));

var _provider = _interopRequireDefault(require("./provider"));
